source/extensions/filters/http/common/compressor/compressor.cc

```diff
@@ -0,0 +1,412 @@
+  decoder_callbacks_ = &callbacks;
+
+  absl::string_view key = compressorRegistryKey();
+  const StreamInfo::FilterStateSharedPtr& filter_state = callbacks.streamInfo().filterState();
+  if (filter_state->hasData<CompressorRegistry>(key)) {
+    CompressorRegistry& registry = filter_state->getDataMutable<CompressorRegistry>(key);
+    registry.filter_configs_.push_back(config_);
+  } else {
+    auto registry_ptr = std::make_unique<CompressorRegistry>();
+    registry_ptr->filter_configs_.push_back(config_);
+    filter_state->setData(key, std::move(registry_ptr),
+                          StreamInfo::FilterState::StateType::Mutable);
+  }
+}
+
+Http::FilterHeadersStatus CompressorFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
+                                                          bool end_stream) {
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isAcceptEncodingAllowed(headers) && isContentTypeAllowed(headers) &&
+      !hasCacheControlNoTransform(headers) && isEtagAllowed(headers) &&
+      isTransferEncodingAllowed(headers) && !headers.ContentEncoding()) {
+    sanitizeEtagHeader(headers);
+    insertVaryHeader(headers);
+    headers.removeContentLength();
+    headers.setContentEncoding(config_->contentEncoding());
+    config_->stats().compressed_.inc();
+    compressor_ = config_->makeCompressor();
+  } else if (!skip_compression_) {
+    skip_compression_ = true;
+    config_->stats().not_compressed_.inc();
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus CompressorFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (!skip_compression_) {
+    config_->stats().total_uncompressed_bytes_.add(data.length());
+    compressor_->compress(data, end_stream ? Compressor::State::Finish : Compressor::State::Flush);
+    config_->stats().total_compressed_bytes_.add(data.length());
+  }
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus CompressorFilter::encodeTrailers(Http::ResponseTrailerMap&) {
+  if (!skip_compression_) {
+    Buffer::OwnedImpl empty_buffer;
+    compressor_->compress(empty_buffer, Compressor::State::Finish);
+    config_->stats().total_compressed_bytes_.add(empty_buffer.length());
+    encoder_callbacks_->addEncodedData(empty_buffer, true);
+  }
+  return Http::FilterTrailersStatus::Continue;
+}
+
+bool CompressorFilter::hasCacheControlNoTransform(Http::ResponseHeaderMap& headers) const {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  if (cache_control) {
+    return StringUtil::caseFindToken(cache_control->value().getStringView(), ",",
+                                     Http::Headers::get().CacheControlValues.NoTransform);
+  }
+
+  return false;
+}
+
+std::unique_ptr<CompressorFilter::EncodingDecision>
+CompressorFilter::chooseEncoding(const Http::ResponseHeaderMap& headers) const {
```

Comment:
This function will need an explicit fuzzer. Also, should any of this be split out into a generic accept-encoding parser in HTTP utilities? Please add a TODO for writing the fuzzer. cc @asraa

Reply:
> Also, should any of this be split out into a generic accept-encoding parser in HTTP utilities?

I'm not sure. Where such utility could be re-used?

Reply:
Anything that parses accept-encoding/q values? It's fine to keep here for now but I could see this easily being reimplemented later. cc @jmarantz